### PR TITLE
Added Paginated GET endpoint

### PR DESCRIPTION
**Pull request by aggKartik**
### Problem

We need to be able to return the most recent queries that a user has made. This list also needs to be paginated so that we don't return every query a user has made. 

### Solution

We already had a working GET endpoint that returned all queries a user made. We had to refactor this so that it accepts a page number and a page size. Also this list had to be ordered by 'added on' so that it would show the most recent queries first. 

### Test

Using postman:

1. Add some entries by hitting the following endpoint:
    
    You can configure the query however you want.

    Endpoint: POST http://localhost:8084/search/userSearch/v1/query
    Body: {"userId": 1, "query": "cola1"}

2. Query the results:

     Endpoint: GET http://localhost:8084/search/userSearch/v1/query
     Query Params: 
                userId : 1
                pageSize : 3
                page: 1

    You can also not include pageSize and page and this endpoint will work.

### CODE REVIEW

File: src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java [hunk 1]
```diff
@@ -25,16 +25,36 @@ public void setServices(UserSearchService categoryService) {
...
+    /**
+     * This endpoint returns a paginated list of queries that a user has made.
+     *
+     * @param id       specify user id to be queried
+     * @param pageSize Optional parameter: specify number of queries per page
+     * @param page     Optional parameter: specify page number of results
+     * @return list of results
+     */
+    @GetMapping(path = "query") //expect page index starting from 1, so subtract one for querying purposes
+    public ResponseEntity<Response<SearchHistoryResponse>> getRecentQueriesForUser(@Valid UserId id,
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java]: Let's change the param to "SearchHistoryRequest" with the object having Long userId, Integer size, and Integer page?

With @NotNull for `userId` and default values for `page` and `size`?

File: src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java [hunk 2]
```diff
@@ -25,16 +25,36 @@ public void setServices(UserSearchService categoryService) {
...
+                                                                                   @RequestParam(required = false) Integer pageSize,
+                                                                                   @RequestParam(required = false) Integer page) {
+        // handle default page size value
+        if (pageSize == null) {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java]: Can be moved inside `SearchHistoryRequest` object

File: src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java [hunk 3]
```diff
@@ -25,16 +25,36 @@ public void setServices(UserSearchService categoryService) {
...
+            pageSize = 10;
+        }
+        // handle default page value
+        if (page == null) {
+            page = 0;
+        } else {
+            page = page - 1;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java]: I think it's fine to have the request accept 0?

File: src/main/java/com/fcgl/madrid/search/payload/response/PaginatedResponse.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package com.fcgl.madrid.search.payload.response;
+
+/**
+ * This class serves as a wrapper for returning a Paginated responses
+ *
+ * @param <T> Response
+ */
+public class PaginatedResponse<T> {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/search/payload/response/PaginatedResponse.java:8]: Good stuff

File: src/main/java/com/fcgl/madrid/search/service/UserSearchService.java [hunk 5]
```diff
@@ -54,9 +60,23 @@ public UserSearchService(UserSearchRepository userSearchRepo) {
...
+     * @return response
+     */
+    public ResponseEntity<Response<SearchHistoryResponse>> searchByUserIdRecent(UserId id, int pageSize, int page) {
+        // Paginate List
+        Pageable pageableRequest = PageRequest.of(page, pageSize);
+        List<UserSearch> searchHistory = userSearchRepo.findByUserIdOrderByAddedOnDesc(id, pageableRequest);
+        // Get total number of records
+        int records = userSearchRepo.countByUserId(id);
+        // Format Response
+        PaginatedResponse paginatedResponse = new PaginatedResponse<>(new SearchHistoryResponse(searchHistory), pageSize, page + 1, records); // to user we want to show non-zero index
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/search/service/UserSearchService.java]: Oh I see. The +1 is for the user. In the app you scroll down to go to the next page. We're using 0 in all the other services though. Maybe the frontend should be in charge of this logic if needed?

**Review by fcgl — CHANGES_REQUESTED** [review]
* Made a few comments
* Will test after update, but code looks like it should work

File: src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java (hunk 1 above)
Comment (aggKartik) [reply src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java]: Sure.

File: src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java (hunk 2 above)
Comment (aggKartik) [reply src/main/java/com/fcgl/madrid/search/controller/UserSearchController.java]: Done.

File: src/main/java/com/fcgl/madrid/search/service/UserSearchService.java (hunk 5 above)
Comment (aggKartik) [reply src/main/java/com/fcgl/madrid/search/service/UserSearchService.java]: I agree, made that change to reflect this.

**Review by fcgl — APPROVED** [review]
* Code looks good ✅ 
* My internet is very slow during the day so unable to test, but I'll trust you on this one
* Will test at night when I update the frontend with the new reponse
* Approved 👍

**Pull request merged**